### ISSUE TITLE
am (Add missing libspec): new item not merged into existing ones

### ISSUE THREAD

**Issue by vemv** [clojure-emacs/clj-refactor.el#406]
Using `am` for refer-ing `<!`, at least in cljs, consistently transitions from this:

```
   [cljs.core.async :refer [>! alts! chan timeout]]
```

to this:

```
   [cljs.core.async :refer [>! alts! chan timeout]]
   [cljs.core.async :refer [<!]]
```

Is there currently a "merge with maybe-existing" functionality?

[my elpa dependencies](https://gist.githubusercontent.com/vemv/ff9e50f6f54a0dea6d2c88a0a38885c4/raw/0a6c6110f7d92382949f7c7d8f31827439a928ed/-) / OSX / Emacs 25.3.1

**Comment by benedekfazekas** [comment]
@expez will confirm or deny my guess but perhaps clean-ns is run after `am`?! so perhaps this is a clean ns issue with core.async?!

**Comment by expez** [comment]
There's a `maybe-clean-ns` call after adding a libspec, but it only runs if `cljr-auto-clean-ns` is truthy.

Is this variable set to `nil` in your emacs, @vemv?

**Comment by vemv** [comment]
Hey there,

I set that and related variables to `t`. All of them were explicitly set to `nil` due to personal preferences.

And yes it works now!

But, cleaning the whole ns as a means to merge things is rather coarse-grained.

For example, cljr's clean-ns removes unused requires, but those requires may be there for their initialization side effects (admittedly a bad practice; but that's how a certain project is architected anyway).

So it's not ideal to intend to just add a refer,  and have in turn valuable requires removed, as a side-effect.

There are also other reasons why cljr-clean-ns is not viable for me.

Hopefully it would be possible to decouple `am` from namespace-cleaning?

Cheers - Victor

**Comment by expez** [comment]
> Hopefully it would be possible to decouple am from namespace-cleaning?

They are decoupled, but that caused you to open an issue when your namespace wasn't cleaned after the libspec was added :p 

There's an issue, I think it's in refactor-nrepl` to add a piece of metadata to a libspec to avoid having it pruned.  There's also a global var to indicate that across all your projects, certain namespaces are known to be side-effecting and should be left alone.  If you find anything missing there, we'd love a PR.

In the meanwhile here are two other solutions:

1. Create a namespace just for these side-effect magic, where you won't be editing code
2. Add a `(require ...)` form below the ns form just for these namespaces

The benefit of both approaches is that it shows that these namespaces are special in some way and you can add a comment or whatever explaining why they are needed even when they appear unused.

Using one of these is a good idea, regardless of any tooling limitations, imo, because some junior developer might remove them manually anyway, because they don't appear to be used :p

**Comment by vemv** [comment]
`am` is coupled to ns-cleaning in that it's not pretty to use one without the other. ('coupling' perhaps being an overly strong term)

I do opt for not using cljr-clean-ns for other two additional reasons:

* I favor placing a newline right after `:require`, that gets undone
* Lines are wrapped around column 80. Tried to get around that back in the day, failed.

https://github.com/gfredericks/how-to-ns does the job a bit more agnostically.

So, a general-purpose fix would be to make `am` try to insert the `<!` (following my original example) in the right place, if it exists.

I can get that this is a slightly unusual need, but still, it's reasonable to consider it an issue in clj-refactor.el ?

It would be either that, or implementing something similar in https://github.com/gfredericks/how-to-ns .

Feel free to close if you don't foresee this getting implemented

**Comment by expez** [comment]
> Feel free to close if you don't foresee this getting implemented

Honestly, I really want to get out of the business of pretty-printing the ns form.  The code we've eventually settled on is ugly, hard to change and extend, and clearly doesn't do everything everyone wants.  I've just resigned myself to 'good enough' and can't remember the last time I spent a second thought on a ns form.

Ideally we'd have library that we could pull in, that supported all dialects and was extensible to such a degree that everyone could be made happy.  Until such a library reveals itself (I tested a few a while back), I'd rather not sink any more time into this particular problem
